File: parser.c

/*
Written by Peter Nelson
Intended to parse json file from Daily Emerald grade data into easily searchable format

Created: 1/18/23
Modified: 2/3/23

Instructions for use:
    copy json file contents to a textfile named gradedata.txt
    delete line 1 and all code at the bottom of the file, leaving only data
    move gradedata.txt to the same directory as parser.c
    compile parser.c with "gcc parser.c" on linux terminal
    execute with "./a.out"
    parsed data will be output to a file named parsedData.txt
*/



#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>


int main(int argc, char* argv[]) {

    //open gradedata file
    FILE *f1 = fopen("gradedata.txt", "r");
    //open file for formatted output
    FILE *f2 = fopen("temp.txt", "w");

    char buff[255];
    char readchar;

    //for every character in gradedata.txt
    while(!feof(f1)) {
        readchar = fgetc(f1);
        //if character is not alphabetic, a digit or a newline, print to temp.txt
        if(isalpha(readchar) != 0 || isdigit(readchar) != 0 || readchar == '\n') {
            fprintf(f2, "%c", readchar);
        }
    }

    //close previous files
    fclose(f1);
    fclose(f2);

    //open temp.txt for reading and parsedData.txt for writing
    FILE *f3 = fopen("temp.txt", "r");
    FILE *f4 = fopen("parsedData.txt", "w");

    char className[10];
    char lastName[30];
    char firstName[30];
    char level[5];

    int storedClass = 0;
    int nameSeparated = 0;

    //for each line of temp.txt
    while(fscanf(f3, "%s", buff) == 1) {
        //if line starts with T, c, or b, then ignore
        if(!(buff[0] == 'T') && !(buff[0] == 'c') && !(buff[0] == 'b')) {
            //if first character of line is capital, must be class name
            if(isupper(buff[0]) != 0) {
                strcpy(className, buff);
                //for each character in the class name, look for first digit
                for(int i = 0; i < strlen(buff); i++) {
                    if(isdigit(buff[i])) {
                        level[0] = buff[i];
                        break;
                    }
                }
                //create level number from above search
                level[1] = '0'; level[2] = '0'; level[3] = '\0';
                storedClass = 1;
            }
            //write out classname and level to parsedData.txt
            if(storedClass == 0 && buff[0] == 'a') {
                fprintf(f4, "%s, ", className);
                fprintf(f4, "%s, ", level);
            }
            nameSeparated = 0;
            //if line starts with i, must be instructor name
            if(buff[0] == 'i') {
                //ignore 'instructor' word, separate name by a comma between last and first name
                for(int i = 10; i < strlen(buff); i++) {
                    if(i > 11 && nameSeparated == 0 && isupper(buff[i]) != 0) {
                        fprintf(f4, "%c ", ',');
                        nameSeparated = 1;
                    }
                    fprintf(f4, "%c", buff[i]);
                }
                fprintf(f4, "%c", '\n');
                storedClass = 0;
            }
            //write out line and level
            else fprintf(f4, "%s, ", buff);
            if(isupper(buff[0]) != 0) {
                fprintf(f4, "%s, ", level);
            }
        }
    }

    //close files
    fclose(f3);
    fclose(f4);
    return 0;
}

File: parser.py
##########################################################
## parser.py by Krishna Patel (kp)
## Created 1/19/23
## Last Modified: 2/2/23
##########################################################

"""
Purpose: This file will look through the JSON file and generate a CSV file usable by the EasyA program. 
================ Revision History ================
File created                         kp    1/19/23
Data parsing implementation          kp    1/19/23
completed                         
Update to exclude middle names of    kp    1/21/23
instructors                  
Update to write to file instead      kp    2/02/23
of to terminal.       
Clean up code                        kp    2/04/23
==================================================

"""
import json
import re


def formatLine(dataEntry, className):
    CSV_line = "" #initialize an empty string
        #all important data elements will be seperated with a comma to stay consistent with a CSV data format
    course = re.split('(\d+)', className) #split course name (ex. CS210) to seperate subject code and class number (ex. [CS, 210])
    CSV_line += course[0] + "," + course[1] + "," #add course subject and number to string
    CSV_line += dataEntry["TERM_DESC"]+ "," #add the term the course was offered to string
    CSV_line +=dataEntry["aprec"]+ "," #add percentage of A's to string
    CSV_line +=dataEntry["bprec"]+ "," #add percentage of B's to string
    CSV_line +=dataEntry["cprec"]+ "," #add percentage of C's to string
    CSV_line +=dataEntry["crn"]+ ","   #add course CRN to string
    CSV_line +=dataEntry["dprec"]+ "," #add percentage of D's to string
    CSV_line +=dataEntry["fprec"]+ "," #add percentage of F's to string
    if(len(dataEntry["instructor"])): #If the instructor name is valid it will be added to the string
        name = dataEntry["instructor"].split(",") #split instructor name at the ","
        CSV_line += name[0] + "," + name[1].split()[0] #add name to string, but split first and middle name at comma 
        #and only add first name to string
    else:
        """
        There was one course that had no instructor assigned, to handle this case the name assigned
        will be NULL for both the first and last name.
        """
        CSV_line += "NULL" + "," +"NULL"

    return CSV_line

def processFile(inputFile, outFile):
    """
    This function will open the input and output file, and will write the parsed data to the output file.
    """
    try:
        JSONfile = open(inputFile, "r") #open JSONfile in read mode and assign filestream to variable "JSONFile"
    except IOError: #if the file does not exist we will exit the program after printing an error
        print("Data file does not exist, please execute program again with real file.")
        exit()
    outputFile = open(outFile, "w") #open output file with filename "parsed_data.csv" in write mode
    #and assign filestream to variable "outputFile"
    data = json.load(JSONfile) #loads all data from JSONFile into a dictionary for easy lookup
    outputFile.write("subject, course number, term, aprec, bprec, cprec, crn, dprec, fprec, last, first\n") #write a CSV header to the output file
    for className in data: #Go through each class in the dataset
        for i in range(len(data[className])):
            """
            The JSON file grouped the same courses into one dictionary index, so we will go through each instance
            to ensure every time a course was offered is recorded.
            """
            dataEntry = data[className][i]
            CSV_line = formatLine(dataEntry, className)
            outputFile.write(CSV_line + "\n") #Write the line of data to the file.
    JSONfile.close() #close the JSON file
    outputFile.close() #close the output file
    return 0

if __name__ == "__main__":
    processFile("gradedata.js", "parsed_data.csv")